src/Features/Core/Portable/ReplaceConditionalWithStatements/AbstractReplaceConditionalWithStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,295 @@
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var generator = SyntaxGenerator.GetGenerator(document);
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        // When we have `a = x ? y : z`, then the type of 'y' and 'z' can influence each other.
+        // If we convert this into:
+        //
+        //   if (x)
+        //     a = y;
+        //   else
+        //     a = z;
+        //
+        // Then we need to preserve that meaning so that 'y' and 'z' have the same type/value, even after the
+        // transformation.
+        var conditionalType = semanticModel.GetTypeInfo(conditionalExpression, cancellationToken).Type;
+
+        syntaxFacts.GetPartsOfAssignmentStatement(assignmentStatement, out var left, out _);
+        syntaxFacts.GetPartsOfConditionalExpression(conditionalExpression, out var condition, out var whenTrue, out var whenFalse);
+        var ifStatement = generator.IfStatement(
+            condition,
+            new[] { generator.AssignmentStatement(left, TryCast(generator, whenTrue, conditionalType)) },
+            new[] { generator.AssignmentStatement(left, TryCast(generator, whenFalse, conditionalType)) }).WithTriviaFrom(assignmentStatement);
+
+        var editor = new SyntaxEditor(root, generator);
+        editor.ReplaceNode(
+            assignmentStatement,
+            ifStatement);
+
+        return document.WithSyntaxRoot(editor.GetChangedRoot());
+    }
+
+    private async Task<Document> ReplaceConditionalExpressionInLocalDeclarationStatementAsync(
+        Document document,
+        TConditionalExpressionSyntax conditionalExpression,
+        TVariableSyntax variable,
+        TLocalDeclarationStatementSyntax localDeclarationStatement,
+        CancellationToken cancellationToken)
+    {
+        var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var generator = SyntaxGenerator.GetGenerator(document);
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        var symbol = (ILocalSymbol)semanticModel.GetRequiredDeclaredSymbol(variable, cancellationToken);
+
+        // When we have `object v = x ? y : z`, then the type of 'y' and 'z' can influence each other.
+        // If we convert this into:
+        //
+        // object v;
+        // if (x)
+        //   v = y;
+        // else
+        //   v = z;
+        //
+        // Then we need to preserve that meaning so that 'y' and 'z' have the same type/value, even after the
+        // transformation.
+        //
+        // Similarly, if we have 'var v', we need to give it a strong type at the declaration point.
+        var conditionalType = semanticModel.GetTypeInfo(conditionalExpression, cancellationToken).Type;
```

Comment:
Is there a scenario where `Type` and `ConvertedType` are different here? It would be good to test such scenario and be clear why we used one and not the other.

Reply:
> Is there a scenario where Type and ConvertedType are different here?

Yes.  that's commonly the case (and tests cover this).  For example, if you have `object o = a ? some_int : some_long`.

The type of hte conditional is `long`, but the converted type is `object`.  We want to the former when working with `some_int` so it loses its int-ness and becomes a long before becoming an object.